File: packages/js/test/helpers/fake-s3.js
// Minimal in-memory S3 for exercising runner mechanics: bucket
// create/delete, object put/get, empty version/upload listings, no locking.
// Port of the Go runner's fakeS3 test helper.

import http from 'node:http'
import { once } from 'node:events'
import { createHash } from 'node:crypto'

export async function startFakeS3 () {
  const buckets = new Map() // name -> Map(key -> Buffer)
  const server = http.createServer((req, res) => {
    const chunks = []
    req.on('data', (c) => chunks.push(c))
    req.on('end', () => handle(req, Buffer.concat(chunks), res, buckets))
  })
  server.listen(0, '127.0.0.1')
  await once(server, 'listening')
  return {
    url: `http://127.0.0.1:${server.address().port}`,
    buckets,
    close: () => new Promise((r) => server.close(r))
  }
}

function handle (req, body, res, buckets) {
  const u = new URL(req.url, 'http://x')
  const parts = u.pathname.replace(/^\/+/, '').split('/')
  const bucket = decodeURIComponent(parts[0] ?? '')
  const key = decodeURIComponent(parts.slice(1).join('/'))
  const objects = buckets.get(bucket)

  const xmlError = (status, code) => {
    res.writeHead(status, { 'content-type': 'application/xml' })
    res.end(req.method === 'HEAD' ? undefined : `<?xml version="1.0"?><Error><Code>${code}</Code><Message>${code}</Message></Error>`)
  }

  if (key === '' && req.method === 'PUT') { // CreateBucket
    buckets.set(bucket, new Map())
    res.writeHead(200).end()
  } else if (key === '' && req.method === 'DELETE') {
    buckets.delete(bucket)
    res.writeHead(204).end()
  } else if (key === '' && u.searchParams.has('uploads')) {
    res.writeHead(200, { 'content-type': 'application/xml' })
    res.end('<?xml version="1.0"?><ListMultipartUploadsResult><IsTruncated>false</IsTruncated></ListMultipartUploadsResult>')
  } else if (key === '' && u.searchParams.has('versions')) {
    let xml = '<?xml version="1.0"?><ListVersionsResult><IsTruncated>false</IsTruncated>'
    for (const k of objects?.keys() ?? []) {
      xml += `<Version><Key>${k}</Key><VersionId>null</VersionId></Version>`
    }
    xml += '</ListVersionsResult>'
    res.writeHead(200, { 'content-type': 'application/xml' }).end(xml)
  } else if (key === '' && u.searchParams.has('object-lock')) {
    xmlError(404, 'ObjectLockConfigurationNotFoundError')
  } else if (key === '' && u.searchParams.has('delete') && req.method === 'POST') {
    for (const m of body.toString().matchAll(/<Key>([^<]*)<\/Key>/g)) objects?.delete(m[1])
    res.writeHead(200, { 'content-type': 'application/xml' })
    res.end('<?xml version="1.0"?><DeleteResult></DeleteResult>')
  } else if (key !== '' && req.method === 'PUT') {
    if (!objects) return xmlError(404, 'NoSuchBucket')
    objects.set(key, body)
    res.writeHead(200, { ETag: `"${createHash('md5').update(body).digest('hex')}"` }).end()
  } else if (key !== '' && (req.method === 'GET' || req.method === 'HEAD')) {
    const data = objects?.get(key)
    if (data === undefined) return xmlError(404, 'NoSuchKey')
    res.writeHead(200, {
      ETag: `"${createHash('md5').update(data).digest('hex')}"`,
      'content-length': String(data.length),
      'content-type': 'application/octet-stream'
    })
    res.end(req.method === 'GET' ? data : undefined)
  } else {
    xmlError(400, 'NotImplemented')
  }
}

File: packages/python/tests/helpers/fake_s3.py
"""Minimal in-memory S3 for exercising runner mechanics: bucket
create/delete, object put/get, empty version/upload listings, no locking.
Port of the JS runner's fake-s3 test helper."""

from __future__ import annotations

import hashlib
import re
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import parse_qs, unquote, urlsplit


class FakeS3:
    def __init__(self) -> None:
        self.buckets: dict[str, dict[str, bytes]] = {}  # name -> key -> bytes
        self.lock = threading.Lock()
        srv = self

        class Handler(BaseHTTPRequestHandler):
            protocol_version = "HTTP/1.1"

            def log_message(self, *args):
                pass

            def _handle(self):
                n = int(self.headers.get("content-length") or 0)
                body = self.rfile.read(n) if n else b""
                status, headers, out = srv._route(self.command, self.path, body)
                self.send_response(status)
                for k, v in headers.items():
                    self.send_header(k, v)
                self.send_header("Content-Length", str(len(out)))
                self.end_headers()
                if self.command != "HEAD":
                    self.wfile.write(out)

            do_GET = do_HEAD = do_PUT = do_POST = do_DELETE = _handle

        self._server = ThreadingHTTPServer(("127.0.0.1", 0), Handler)
        self.url = f"http://127.0.0.1:{self._server.server_address[1]}"
        self._thread = threading.Thread(target=self._server.serve_forever, daemon=True)
        self._thread.start()

    def close(self) -> None:
        self._server.shutdown()
        self._server.server_close()
        self._thread.join()

    def _route(self, method: str, path: str, body: bytes):
        u = urlsplit(path)
        parts = u.path.lstrip("/").split("/")
        bucket = unquote(parts[0]) if parts else ""
        key = unquote("/".join(parts[1:]))
        q = parse_qs(u.query, keep_blank_values=True)
        xml = {"content-type": "application/xml"}

        def error(status, code):
            return status, xml, f'<?xml version="1.0"?><Error><Code>{code}</Code><Message>{code}</Message></Error>'.encode()

        with self.lock:
            objects = self.buckets.get(bucket)
            if key == "" and method == "PUT":  # CreateBucket
                self.buckets[bucket] = {}
                return 200, {}, b""
            if key == "" and method == "DELETE":
                self.buckets.pop(bucket, None)
                return 204, {}, b""
            if key == "" and "uploads" in q:
                return 200, xml, b'<?xml version="1.0"?><ListMultipartUploadsResult><IsTruncated>false</IsTruncated></ListMultipartUploadsResult>'
            if key == "" and "versions" in q:
                out = '<?xml version="1.0"?><ListVersionsResult><IsTruncated>false</IsTruncated>'
                for k in (objects or {}):
                    out += f"<Version><Key>{k}</Key><VersionId>null</VersionId></Version>"
                out += "</ListVersionsResult>"
                return 200, xml, out.encode()
            if key == "" and "object-lock" in q:
                return error(404, "ObjectLockConfigurationNotFoundError")
            if key == "" and "delete" in q and method == "POST":
                for m in re.finditer(r"<Key>([^<]*)</Key>", body.decode("utf-8", "replace")):
                    if objects is not None:
                        objects.pop(m.group(1), None)
                return 200, xml, b'<?xml version="1.0"?><DeleteResult></DeleteResult>'
            if key != "" and method == "PUT":
                if objects is None:
                    return error(404, "NoSuchBucket")
                objects[key] = body
                return 200, {"ETag": f'"{hashlib.md5(body).hexdigest()}"'}, b""
            if key != "" and method in ("GET", "HEAD"):
                data = (objects or {}).get(key)
                if data is None:
                    return error(404, "NoSuchKey")
                headers = {"ETag": f'"{hashlib.md5(data).hexdigest()}"', "content-type": "application/octet-stream"}
                return 200, headers, data
            return error(400, "NotImplemented")
